### DISCUSSION TITLE
Plugin is unable process resource links (Docusaurus process them anyway)

### DISCUSSION THREAD

**Discussion by juhamust** [facebook/docusaurus#9041] [General]
I would like to replace links to resources with absolute links

## Document *before processing*

```markdown
Please check [package.json](../../frontend/package.json) 
for dependencies.
```

## Document *after processing* (wanted outcome)

```markdown
Please check [package.json](https://github.com/username/projectx/blob/development/frontend/package.json) 
for dependencies.
```

> NOTE: Even if the plugin processes the links the Docusaurus includes the files in build directory, which is _not_ wanted.

## Plugin code

```javascript
const path = require('path');
const unistReady = import('unist-util-visit');

const DEFAULT_LINK_PREFIX = 'https://github.com/username/projectx/blob/development';
const DEFAULT_RESOURCE_EXTENSIONS = [
  '.json',
  '.ts',
  '.tsx',
  '.js',
  '.jsx',
];

/**
 * Remark plugin that makes it easy to create links to files in GitHub
 * See: https://github.com/syntax-tree/mdast
 * @param linkPrefix baseURL like https://github.com/org/repo/blob/branch
 * @returns
 */
const gitLinkPlugin = ({ linkPrefix = DEFAULT_LINK_PREFIX, resourceExtensions = DEFAULT_RESOURCE_EXTENSIONS }) => {

  const getMarkdownLink = ({ title = null, filePath }) => {
    const link = `${linkPrefix}/${filePath}`;

    return {
      type: 'link',
      url: link,
      title: link,
      children: [{type: 'text', value: title ?? link }],
    }
  }

  const transformer = async (ast) => {
    const unist = await unistReady;

    // Find and replace the following links with absolute GitHub links:
    //
    // - [package.json](../../packages/frontend/package.json)
    //
    unist.visit(ast, 'link', (node) => {
      const fileExtension = path.extname(node.url);
      if (resourceExtensions.includes(fileExtension)) {
        const filePath = node.url.replace(/^[\.\/]*/, '');
        const markdownLink = getMarkdownLink({ filePath });
        node.url = markdownLink.url;
        node.title = markdownLink.title;
      }
    });
  };
  return transformer;
};

module.exports = gitLinkPlugin;
```

**Comment by slorber** [comment]
yes this is a problem, the legacy way we process links is based on the string content of the doc instead of using the AST, so adding a remark plugin runs after that linkification. CF my comment here: https://github.com/facebook/docusaurus/pull/8927#issuecomment-1525937829 - I think you can open a bug report, we should support this better

Try this workaround: `[package.json](pathname://../../frontend/package.json)`, the link should be left untouched and you can process it with remark.

**Reply by juhamust** [reply]
Thanks for the input. The issue with the [described solution](https://docusaurus.io/docs/markdown-features/assets#static-assets) is that then IDE is no longer able to check the link validity or navigate to the file 😞

**Reply by juhamust** [reply]
Will create a ticket about this.

**Reply by slorber** [reply]
Hmmm, thinking about it, the "linkify" code is only responsible to the transformation of links ending with `.mdx?`

I think you should be able to achieve your need by using option `beforeDefaultRemarkPlugins` instead of `remarkPlugins`

**Reply by juhamust** [reply]
Thanks for the tip - will try that one as well 👍

**Reply by slorber** [reply]
Great, let me know if it works. Normally your plugin will run before the Docusaurus one with that option.

Finally unrelated, but I created the proposal for the other linkify change I suggested we should do here: https://github.com/facebook/docusaurus/issues/9048
